packages/internal/src/callback.js

```diff
@@ -155,3 +183,132 @@ export const isCallback = callback => {
   );
 };
 harden(isCallback);
+
+/**
+ * Prepare an attenuator class whose methods can be redirected via callbacks.
+ *
+ * @template {PropertyKey} M
+ * @param {import('@agoric/zone').Zone} zone The zone in which to allocate attenuators.
+ * @param {M[]} methodNames Methods to forward.
+ * @param {object} opts
+ * @param {InterfaceGuard} [opts.interfaceGuard] An interface guard for the
+ * new attenuator.
+ * @param {string} [opts.tag] A tag for the new attenuator exoClass.
+ */
+export const prepareAttenuator = (
+  zone,
+  methodNames,
+  { interfaceGuard, tag = 'Attenuator' } = {},
+) => {
+  /**
+   * @typedef {(this: any, ...args: unknown[]) => any} Method
+   * @typedef {{ [K in M]: Method }} Methods
+   * @typedef {{ [K in M]?: Callback<any> | null}} Overrides
+   */
+  const methods = fromEntries(
+    methodNames.map(key => {
+      // Only allow the `PropertyKey` type for the target method key.
+      if (!isPropertyKey(key)) {
+        throw Fail`key ${q(key)} is not a PropertyKey`;
+      }
+
+      const m = /** @type {Methods} */ ({
+        // Explicitly use concise method syntax to preserve `this` but prevent
+        // constructor behavior.
+        /** @type {Method} */
+        [key](...args) {
+          // Support both synchronous and async callbacks.
+          const cb = this.state.cbs[key];
+          if (!cb) {
+            const err = assert.error(
+              `unimplemented ${q(tag)} method ${q(key)}`,
+            );
+            if (this.state.isSync) {
+              throw err;
+            }
+            return Promise.reject(err);
+          }
+          if (cb.isSync) {
+            return callSync(cb, ...args);
+          }
+          return callE(cb, ...args);
+        },
+      })[key];
+      return /** @type {const} */ ([key, m]);
+    }),
+  );
+
+  const methodKeys = /** @type {M[]} */ (ownKeys(methods));
+
+  /**
+   * Create an exo object whose behavior is composed from a default target
+   * and/or individual method override callbacks.
+   *
+   * @param {object} opts
+   * @param {unknown} [opts.target] The target for any methods that
+   * weren't specified in `opts.overrides`.
```

Comment:
What is the use case for making target optional (e.g., `makeAttenuator({ target: undefined, overrides })`)?

Reply:
If you want to compose an object entirely out of callbacks, you don't need a default target.

Reply:
Should we encourage use of `makeAttenuator` for composing an object from callbacks? I would expect not, as opposed to e.g.
```js
const makeInstance = zone.exoClass(tag, interfaceGuard, initEmpty, { ...callbacks });
```

Reply:
(Aside from legacy...)

Now that we have attenuators, when would we ever use callbacks? AFAICT, attenuators are just better. If so, perhaps we should try to phase out callbacks over time.

Reply:
I don't see how attenuators supplant callbacks. Thanks to callbacks you can compose a new object of a given shape from different existing objects.